File: data-structure/Set/BSTSet.ts
import BST from '../BinarySearchTree/BST';

interface ISet<E> {
  add(e: E): void;
  contains(e: E): boolean;
  remove(e: E): void;
  getSize(): number;
  isEmpty(): boolean;
}

class BSTSet<E> implements ISet<E> {
  private bst: BST<E> = new BST<E>();

  getSize() {
    return this.bst.getSize();
  }

  isEmpty() {
    return this.bst.isEmpty();
  }

  add(e: E): void {
    this.bst.add(this.bst.root, e);
  }

  contains(e: E): boolean {
    return this.bst.contains(e);
  }

  remove(e: E): void {
    this.bst.remove(e);
  }
}

File: DataStructureJava/BSTSet.java
public class BSTSet<E extends Comparable<E>> implements Set<E> {
    private BST<E> bst;

    public BSTSet() {
        bst = new BST<>();
    }

    // 时间复杂度O(1)
    @Override
    public int getSize() {
        return bst.size();
    }

    // 时间复杂度O(1)
    @Override
    public boolean isEmpty() {
        return bst.isEmpty();
    }

    // 时间复杂度O(logN)
    @Override
    public void add(E e) {
        bst.add(e);
    }

    // 时间复杂度O(logN)
    @Override
    public boolean contains(E e) {
        return bst.contains(e);
    }

    // 时间复杂度O(logN)
    @Override
    public void remove(E e) {
        bst.remove(e);
    }
}
